src/mono/mono/utils/mono-codeman.c

```diff
@@ -668,9 +669,11 @@ mono_codeman_enable_write (void)
 		mono_native_tls_set_value (write_level_tls_id, GINT_TO_POINTER (level));
 		pthread_jit_write_protect_np (0);
```

Comment:
This isn't a part of this diff, but: is the Apple-supplied definition of `pthread_jit_write_protect_np` defined with `__attribute__((weak_import))`? If not, do we expect that Mono built on a macOS 11+ amd64 machine will fail during symbol resolution on older OS versions?

Reply:
hm. good catch.

```
__attribute__((availability(macos,introduced=11.0)))
__attribute__((availability(ios,unavailable))) __attribute__((availability(tvos,unavailable))) __attribute__((availability(watchos,unavailable))) __attribute__((availability(driverkit,unavailable)))
void pthread_jit_write_protect_np(int enabled);
```

```
$ nm foo
0000000100008008 d __dyld_private
0000000100000000 T __mh_execute_header
0000000100003f80 T _main
                 U _pthread_jit_write_protect_np
                 U dyld_stub_binder
```

Doesn't look weak.  So yea we should probably not build this on amd64 at all.

Reply:
Hmm... actually if I compile my toy with `-mmacosx-version-min=10.5` then  `nm -m` is better:

```
$ nm -m foo
0000000100004110 (__DATA,__bss) non-external _AvailabilityVersionCheck
                 (undefined) external _CFBundleGetVersionNumber (from CoreFoundation)
00000001000040f0 (__DATA,__bss) non-external _CompatibilityDispatchOnceCounter
0000000100004108 (__DATA,__bss) non-external _DispatchOnceCounter
00000001000040f8 (__DATA,__bss) non-external _GlobalMajor
00000001000040fc (__DATA,__bss) non-external _GlobalMinor
0000000100004100 (__DATA,__bss) non-external _GlobalSubminor
00000001000040d0 (__DATA,__common) external _NXArgc
00000001000040d8 (__DATA,__common) external _NXArgv
                 (undefined) external ___assert_rtn (from libSystem)
0000000100003360 (__TEXT,__text) non-external (was a private external) ___clang_at_available_requires_core_foundation_framework
0000000100003370 (__TEXT,__text) non-external (was a private external) ___isOSVersionAtLeast
0000000100003520 (__TEXT,__text) non-external (was a private external) ___isPlatformOrVariantPlatformVersionAtLeast
0000000100003a40 (__TEXT,__text) non-external ___isPlatformOrVariantPlatformVersionAtLeast.cold.1
0000000100003400 (__TEXT,__text) non-external (was a private external) ___isPlatformVersionAtLeast
00000001000040e8 (__DATA,__common) external ___progname
                 (undefined) external ___stack_chk_fail (from libSystem)
                 (undefined) external ___stack_chk_guard (from libSystem)
00000001000032f0 (__TEXT,__text) non-external (was a private external) __dyld_func_lookup
0000000100003700 (__TEXT,__text) non-external __initializeAvailabilityCheck
0000000100000000 (absolute) [referenced dynamically] external __mh_execute_header
00000001000033f0 (__TEXT,__text) non-external _compatibilityInitializeAvailabilityCheck
                 (undefined) weak external _dispatch_once_f (from libSystem)
                 (undefined) external _dlsym (from libSystem)
00000001000040e0 (__DATA,__common) external _environ
                 (undefined) external _exit (from libSystem)
                 (undefined) external _fclose (from libSystem)
                 (undefined) external _fopen (from libSystem)
                 (undefined) external _fread (from libSystem)
                 (undefined) external _free (from libSystem)
                 (undefined) external _fseek (from libSystem)
                 (undefined) external _ftell (from libSystem)
                 (undefined) weak external _gaz (from libgaz)
0000000100003510 (__TEXT,__text) non-external _initializeAvailabilityCheck
0000000100003300 (__TEXT,__text) external _main
                 (undefined) external _malloc (from libSystem)
                 (undefined) weak external _pthread_jit_write_protect_np (from libSystem)
                 (undefined) external _rewind (from libSystem)
                 (undefined) external _sscanf (from libSystem)
00000001000032dc (__TEXT,__text) non-external (was a private external) dyld_stub_binding_helper
00000001000032a0 (__TEXT,__text) external start
```

in particular `(undefined) weak external _pthread_jit_write_protect_np (from libSystem)`

So as long as we set `-mmacosx-version-min=` we're okay

Reply:
Ah, __attribute__((availability)) does this. It's even documented here: https://clang.llvm.org/docs/AttributeReference.html